### DISCUSSION TITLE
Abstract away base Docker images without C-PAC

### DISCUSSION THREAD

**Discussion by shnizzedy** [FCP-INDI/C-PAC#1455] [Ideas]
Building a Docker image takes awhile ([~½ hour on CircleCI usually](https://app.circleci.com/pipelines/github/FCP-INDI/C-PAC)). Building AFNI from source instead of from the latest binaries (#1454) [nearly doubles that](https://app.circleci.com/pipelines/github/FCP-INDI/C-PAC/1444/workflows/36e67c94-03d0-4f3d-84b3-9008a8eeed83/jobs/5600).

Most of the time, all we're _really_ changing is in this little bit at the end:

https://github.com/FCP-INDI/C-PAC/blob/55ae58c004307294dae41ba4a60120ca444bf465/Dockerfile#L216-L226

I think we could create a `base.Dockerfile` that includes everything in our current Dockerfile except the above, build it with a GitHub Action only when that file changes and [store it here](https://github.com/orgs/FCP-INDI/packages?repo_name=C-PAC), and have the regular `Dockerfile` build from that image instead of https://github.com/FCP-INDI/C-PAC/blob/55ae58c004307294dae41ba4a60120ca444bf465/Dockerfile#L2

I've done this in branches when trying particular versions of dependencies, like https://github.com/FCP-INDI/C-PAC/blob/c17333641aad3d47b54b90c405e23430d340c872/Dockerfile#L1-L2

but I think it would make sense to do this for all of the dependencies, to save that build time on Circle.

Related #788

**Comment by shnizzedy** [comment]
Also, I think we could create an image with regression test outputs, update that image with each release, be clever with `docker-compose`, and really streamline regression testing.

**Reply by sgiavasis** [reply]
😮  yes

**Reply by sgiavasis** [reply]
Is there something I can put together or clarify that can help this along?

**Reply by shnizzedy** [reply]
For the multistage building images, I have a pretty good idea of what to do. For the regression test container, I haven't thought it all the way out yet, so if you get to it before I do, you're welcome to take the reigns. It's one of those things that I expect will take some time to design and build and then save time once it exists

**Reply by sgiavasis** [reply]
Now that the node blocks are modular and we can more easily run them standalone, I am picturing running correlations for regression testing on each one individually - but still keeping some of the main preconfigs for integration testing instead.

Basically, this would remove the necessity for those big regtest-N configs I created that attempt to test almost everything. This would make things leaner.

**Reply by shnizzedy** [reply]
One-by-one regression tests will be fantastic! Way easier to spot and diagnose issues.
